src/rgw/driver/rados/rgw_notify.cc

```diff
@@ -679,16 +699,50 @@ class Manager : public DoutPrefixProvider {
         queue_gc.clear();
       }
     }
+    for (const auto& queue_name : owned_queues) {
+      const auto ret = unlock_queue(queue_name, rados_ioctx);
+      if (ret < 0) {
+        ldpp_dout(this, 5) << "WARNING: failed to unlock queue: " << queue_name << ". ownership would still move if not renewed" << dendl;
+      }
+    }
+    ldpp_dout(this, 5) << "INFO: manager stopped. done processing for queues" << dendl;
   }
 
 public:
 
   ~Manager() {
-    work_guard.reset();
+  }
+
+  void stop() {
+    shutdown = true;
     io_context.stop();
+    work_guard.reset();
```

Comment:
@cbodley if i'm not stopping the io_context, i can get a a crash if the async `unlock()` call completes after the manager already destroyed.
the other option, is to call the sync version of `unlock()` (as i did before), but if we have many queues, this would be an issue.
so, i guess that i would risk that sometimes the unlock is being called and sometimes it does not (i will eventually get the lock, once the lock timer expires).

Reply:
> @cbodley if i'm not stopping the io_context, i can get a a crash if the async `unlock()` call completes after the manager already destroyed.

if you don't call `io_context::stop()`, then any threads calling `io_context.run()` would continue run until all of their coroutines finish. that means the calls to `thread::join()` just below will block until that happens. so i don't think there's a risk of racing with Manager's destruction

Reply:
i tried commenting out this `io_context.stop()` line and the cleanup/shutdown mostly works:
```
2024-05-02T15:28:32.534-0400 7f34c54006c0  5 rgw notify: INFO: manager stopped. done processing for queue: RGW11111111111111111:Topic
2024-05-02T15:28:32.534-0400 7f34c54006c0  1 -- 192.168.245.130:0/802693616 --> [v2:192.168.245.130:6800/3518936173,v1:192.168.245.130:6801/3518936173] -- osd_op(unknown.0.0:3088 2.0 2:0e9ddf52:notif::RGW11111111111111111%3aTopic:head [stat,call lock.unlock in=71b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e13) -- 0x562f2d2eaa80 con 0x562f2ad66480
2024-05-02T15:28:32.536-0400 7f350aa006c0  1 -- 192.168.245.130:0/802693616 <== osd.0 v2:192.168.245.130:6800/3518936173 3090 ==== osd_op_reply(3088 RGW11111111111111111:Topic [stat,call] v13'427 uv427 ondisk = 0) ==== 212+0+0 (crc 0 0 0) 0x562f2ae58d80 con 0x562f2ad66480
2024-05-02T15:28:32.537-0400 7f34c54006c0 10 rgw notify: INFO: queue: RGW11111111111111111:Topic not locked (ownership can move)
2024-05-02T15:28:32.537-0400 7f34c54006c0 10 rgw notify: INFO: queue: RGW11111111111111111:Topic marked for removal
2024-05-02T15:28:33.406-0400 7f34c54006c0  1 -- 192.168.245.130:0/802693616 --> [v2:192.168.245.130:6800/3518936173,v1:192.168.245.130:6801/3518936173] -- osd_op(unknown.0.0:3089 2.0 2:6fe6cc47:notif::queues_list_object:head [omap-get-keys in=12b] snapc 0=[] ondisk+read+known_if_redirected+supports_pool_eio e13) -- 0x562f2d273500 con 0x562f2ad66480
2024-05-02T15:28:33.406-0400 7f350aa006c0  1 -- 192.168.245.130:0/802693616 <== osd.0 v2:192.168.245.130:6800/3518936173 3091 ==== osd_op_reply(3089 queues_list_object [omap-get-keys out=35b] v0'0 uv411 ondisk = 0) ==== 162+0+35 (crc 0 0 0) 0x562f2ae58d80 con 0x562f2ad66480
2024-05-02T15:28:33.406-0400 7f34c54006c0  1 -- 192.168.245.130:0/802693616 --> [v2:192.168.245.130:6800/3518936173,v1:192.168.245.130:6801/3518936173] -- osd_op(unknown.0.0:3090 2.0 2:0e9ddf52:notif::RGW11111111111111111%3aTopic:head [stat,call lock.lock in=87b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e13) -- 0x562f2d273880 con 0x562f2ad66480
2024-05-02T15:28:33.409-0400 7f350aa006c0  1 -- 192.168.245.130:0/802693616 <== osd.0 v2:192.168.245.130:6800/3518936173 3092 ==== osd_op_reply(3090 RGW11111111111111111:Topic [stat,call] v13'428 uv428 ondisk = 0) ==== 212+0+0 (crc 0 0 0) 0x562f2ae58d80 con 0x562f2ad66480
2024-05-02T15:28:33.409-0400 7f34c54006c0 20 rgw notify: INFO: queue: RGW11111111111111111:Topic ownership (lock) renewed
2024-05-02T15:28:33.409-0400 7f34c54006c0 10 rgw notify: INFO: queue: RGW11111111111111111:Topic was removed
2024-05-02T15:28:33.409-0400 7f34c54006c0  5 rgw notify: INFO: manager stopped. done processing for queues
```
the only strange part is that `process_queues()` raced to lock the queue again after `process_queue()` unlocked/returned

Reply:
tried testing locally without `io_context.stop()` when more than one RGW exists:
```
RGW_PER_ZONE=2 ../src/test/rgw/test-rgw-multisite.sh 1
```
but the tests (`data_path_v2_test`) got stuck in "migrating" state.
with the above call the tests are running fine.

Reply:
can you tell which coroutine gets stuck on shutdown?

Reply:
trying with these steps:
```
$ MON=1 OSD=1 RGW=0 MDS=0 MGR=0 RGW_PER_ZONE=2 ../src/test/rgw/test-rgw-multisite.sh 1
$ bin/radosgw-admin -c run/c1/ceph.conf user create --uid testid --access-key 0555b35654ad1656d804 --secret h7GhxuBLTrlhVUyxSPUKUV8r/2EI4ngqJxD7iBdBYLhwluN30JaT3Q== --display-name 'M. Tester' --email tester@ceph.com
$ BNTESTS_CONF=../work/test_bn.conf python -m nose -s ../src/test/rgw/bucket_notification/test_bn.py -v -a data_path_v2_test
```
i see both rgws crashing on Manager shutdown:
```
    -5> 2024-05-03T13:55:17.483-0400 7fd99c4006c0  5 rgw notify: INFO: manager stopped. done processing for queue: :vxhzxs-1_topic
    -4> 2024-05-03T13:55:17.483-0400 7fd99c4006c0  1 -- 192.168.245.130:0/2176396696 --> [v2:192.168.245.130:6800/2282501434,v1:192.168.245.130:6801/2282501434] -- osd_op(unknown.0.0:722 2.0 2:071c9d28:notif::%3avxhzxs-1_topic:head [stat,call lock.unlock in=60b] snapc 0=[] ondisk+write+known_if_redirected+supports_pool_eio e19) -- 0x563fe8e8b880 con 0x563fe0b80900
    -3> 2024-05-03T13:55:17.484-0400 7fd99c4006c0  5 rgw notify: INFO: manager stopped. done processing for queues
    -2> 2024-05-03T13:55:17.485-0400 7fdb102006c0  1 -- 192.168.245.130:0/2176396696 <== osd.0 v2:192.168.245.130:6800/2282501434 731 ==== osd_op_reply(722 :vxhzxs-1_topic [stat,call] v19'1633 uv1633 ondisk = 0) ==== 201+0+0 (crc 0 0 0) 0x563fe592ed80 con 0x563fe0b80900
    -1> 2024-05-03T13:55:17.485-0400 7fd99c4006c0 10 rgw notify: INFO: queue: :vxhzxs-1_topic not locked (ownership can move)
     0> 2024-05-03T13:55:17.485-0400 7fd99c4006c0 -1 *** Caught signal (Segmentation fault) **
 in thread 7fd99c4006c0 thread_name:notif-worker0

 ceph version Development (no_version) squid (dev)
 1: /home/cbodley/ceph/build/bin/radosgw(+0x13bf562) [0x563fde456562]
 2: /lib64/libc.so.6(+0x3e9a0) [0x7fdb1905c9a0]
 3: pthread_mutex_lock()
 4: (std::mutex::lock()+0x9) [0x563fddc0eb7d]
 5: (rgw::notify::Manager::process_queues(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)::{lambda(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)#1}::operator()(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >) const+0x2ab) [0x563fde03c9fb]
 6: (spawn::detail::spawn_helper<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0ul> > >, rgw::notify::Manager::process_queues(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)::{lambda(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)#1}, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits> >::operator()()::{lambda(boost::context::continuation&&)#1}::operator()(boost::context::continuation&&) const+0xc9) [0x563fde03cd53]  
 7: (boost::context::detail::record<boost::context::continuation, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits>, spawn::detail::spawn_helper<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0ul> > >, rgw::notify::Manager::process_queues(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)::{lambda(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)#6}, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits> >::operator()()::{lambda(boost::context::continuation&&)#1}>::run(void*)+0x2e) [0x563fde03cef4]
 8: (void boost::context::detail::context_entry<boost::context::detail::record<boost::context::continuation, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits>, spawn::detail::spawn_helper<boost::asio::executor_binder<void (*)(), boost::asio::strand<boost::asio::io_context::basic_executor_type<std::allocator<void>, 0ul> > >, rgw::notify::Manager::process_queues(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)::{lambda(spawn::basic_yield_context<boost::asio::executor_binder<void (*)(), boost::asio::any_io_executor> >)#1}, boost::context::basic_protected_fixedsize_stack<boost::context::stack_traits> >::operator()()::{lambda(boost::context::continuation&&)#1}> >(boost::context::detail::transfer_t)+0x68) [0x563fde03cfd8]
 9: make_fcontext()
```

Reply:
that's because `queue_gc_lock` (and `queue_gc`) are variables on the stack of `process_queues()`, which can return before the `process_queue()` coroutines return

on shutdown, maybe `process_queues()` should keep looping until `owned_queues` is empty?

Reply:
if this gets too complicated, we might just revert the shutdown/unlock changes and leave io_context.stop() there for now